# Python workshop 2
## Lists, loops, and libraries

*Summer Science Program in Astrophysics*

---

## *Workshop 1: A quick review*

You can use this as a reference if you need to remember anything from the first workshop

### **Variables & Printing**

Python variables can store any type of data.

There are two main ways that Python represents numbers: `int` and `float` An `int` is an integer, i.e. a positive or negative number with no decimal component. A `float` (or "floating point number") is a number that CAN have a fractional or decimal component.
* In general, any number written with a decimal point (e.g. `0.1234, 6.0`) is a float, while a number without one (e.g. `6`, `-25`) is an int.
* You can also create floats using scientific notation. `1.52e6` represents $1,520,000$

You can use `+`, `-`, `*`, and `/` to add, subtract, multiply, and divide. Other Python math operators include
* `**`: Exponentiation. `2**4 -> 16`
* `//`: "Integer division", which divides and ignores the remainder. `16 // 3 -> 5`
* `%`: "Modulus", which returns ONLY the remainder. `16 % 3 -> 1`

To represent **text**, Python uses a variable type called "string" (or just `str`), which you can create using quotation marks. You can use 'single quotes' (`' '`) or "double quotes" (`" "`) to create strings, as long as they match.
* `'This is a string!'`
* `"This is also a string, but with double quotes!"`

The **print** function displays values or text. You can print the value of any variable by typing `print(variable)`. You can even print multiple things: `print("The value of variable is", variable)`

In [7]:
# -- EXAMPLE CODE: Variables and printing --

# Computes the area of a triangle given a base and a height

triangle_base = 8      # This is an int
triangle_height = 4.5  # This is a float

triangle_area = 0.5 * triangle_base * triangle_height

print('Triangle area:', triangle_area)

print('Type of each variable above:')
print(8, type(8))
print(4.5, type(4.5))
print('Some text!', type('Some text!'))

Triangle area: 18.0
Type of each variable above:
8 <class 'int'>
4.5 <class 'float'>
Some text! <class 'str'>


### **True, False, and Conditionals** (a.k.a. "if statements")

When you compare two values, Python outputs a "boolean", a value that's either `True` or `False`. Example:
```
print(2 < 5)
-> True
```

You can compare values in a number of ways:
* `a == b`  Equal 
* `a != b`  Not equal
* `a < b `  Strictly less than
* `a <= b`  Less than or equal to
* `a > b `  Strictly greater than
* `a >= b`  Greater than or equal to

You can also use the key words `and`, `or`, and `not` to chain together comparisons:
```
(a > 0) and (a != 10)
```

An **if statement** will only run a block of code if a condition is `True`. If it's `False`, the code is skipped. It can be followed by an `elif` ("else if") statement or an `else` statement, which will only run if the statements above them did not:
```
if [CONDITION]:
    BLOCK OF CODE
    ...
elif [CONDITION]:
    DIFFERENT CODE
    ...
else:
    THIS CODE RUNS IF NONE OF THE CONDITIONS ABOVE WERE TRUE
    ...
```

Each `if` (or `elif` or `else`) statement MUST end with a colon `:`, and it must be followed up by a block of code that's indented.

In [8]:
# -- EXAMPLE CODE: True/False and conditionals --

# Determines which of "a", "b", or "c" is largest

a = 3
b = 6
c = 9

if (a > b) and (a > c):
    print('a is largest')
    print('Amazing for a!')
elif (b > c):
    print('b is largest')
    print('Bully for b!')
else:
    print('c is largest')
    print('Celebrate c!')
print("(This line is not indented, so it's not part of the if statement)")

c is largest
Celebrate c!
(This line is not indented, so it's not part of the if statement)


### **Functions**

A "function" is a snippet of code that takes in a set of **arguments**, then **returns** a result (with the keyword `return`).

To **define** a function:
```
def add_args(arg1, arg2):
    # Returns the sum of arg1 and arg2
    return arg1 + arg2
```
(When a function `returns` a value, it immediately stops running.)

To **call** a function:
```
add_args(12, 4)
-> 16
```

Throughout the program, when we ask you to write code, we'll almost always ask you to write a function or set of functions. They are **very** useful for keeping your code organized.

You can call functions from anywhere, even from within other functions.

In [9]:
# -- EXAMPLE CODE: Functions --

def get_larger(a, b):
    # Given two numbers "a" and "b", returns the larger one
    if a > b:
        return a
    return b

def get_largest_of_3(a, b, c):
    # Given THREE numbers "a", "b", and "c", returns the largest one
    return get_larger(get_larger(a, b), c)

get_largest_of_3(10, 20, -2)

20

<br>

---

# What's the goal of this workshop?

The Method of Gauss is an *iterative* method. It starts by taking a guess at an orbit, then improving it repeatedly until the orbit changes very little with each improvement. As you run it, you'll want to hold onto a record of each orbital guess so that you can see how the orbit improved over time. Unfortunately, you won't know in advance how many steps the Method of Gauss will require to converge on a solution.

The last workshop covered most of the "Python fundamentals" that you'll need to determine your asteroid's orbit. However, to perform orbit determination, you'll need tools that can repeat steps a *variable* number of times, and that can store data of a *variable* size. The tools you've seen so far can't do these things.

This workshop will introduce you to the tools you'll need, which are **lists** for storing data and **loops** for managing your code. You'll need these skills for the first code problem set to solve problems like:
* Compute the dot product of any two vectors (regardless of their length)
* Sort a list of values from smallest to largest

In addition, this workshop will teach you how to keep your code organized using **library** files. In a few moments, you'll start your own "orbit determination library" that you'll continue to grow throughout the program.

#### TABLE OF CONTENTS

1. **Creating your own library**
2. **Loops**
    * 2.1) "For" loops
    * 2.2) Helpful ways to use loops
    * 2.3) "While" loops
3. **Lists**
    * 3.1) What is a list?
    * 3.2) Lists and loops
    * 3.3) Adding values to a list
    * 3.4) "Slicing" a list
    * 3.5) Beware! Lists are mutable

<br>

---

# 1) 📚 Creating your own library 📚

**This section covers:** How to start your own "OD Library" file, and how to "import" its code from another file.

Putting all of your code into one big a Jupyter notebook would get unwieldy very quickly! To build your OD code, we need a way to:
* Divide up your code over multiple files
* Load code from these files as it's needed

Your first task today: Start your **Orbit Determination Library**, or "odlib".
* This will be like a text file, but containing only Python code
* The name ending in `.py`.
* Your computer will recognize this as a Python code file.
* Over the course of the program, you'll gradually build this library up by adding useful functions to it. These will help you perform the Method of Gauss on your asteroid.

From here forward, each coding assignment during the program will ask you to hand in (at least) one `.py` file.

### **Step 1) Create your library file**

* Press the blue "[+]" button (or press ctrl-shift-L) to start the "Launcher"
* Create a new file called "odlib_[YOUR_LAST_NAME].py"

### **Step 2) Add some of your functions to this file**

Open your "Python workshop 1" file. Copy each of the functions `hms_to_deg` and `dms_to_deg` into your new "odlib" file. (If they're not complete, copy what you have.)

### **Step 3) Import your odlib file to this notebook**

To do so, we need to use the "import" keyword.

Example: Let's say you have a ".py" file called `library_name.py` which contains a function, `example_function`. You could run this function from a different file with the following code:
```
import library_name as lib

lib.example_function()
```
After running this code, you could run any function defined in `library_name` by typing `lib.[FUNCTION_NAME]`

#### **To complete step 3, fill in the cell below to import and run your functions `hms_to_deg` and `dms_to_deg`.**

In [10]:
import sys
import os
sys.path.append(os.path.abspath(".."))

In [11]:
# Change this to match your file's name:
import odlib_agrawal as od

# Vega's RA & Dec: RA = 18h 36m 56s | Dec = +38° 47′ 1″ 
# Use your code to convert these to decimal degrees!
print("Converted RA = ", od.hms_to_deg(18, 36, 56))
print("Converted Dec = ", od.dms_to_deg(True, 38, 47, 1))

Converted RA =  279.23333333333335
Converted Dec =  38.78361111111111


### **Step 4) Add a "main" function**

As a final step, add a new function to your "odlib" file:
```
def main():
    ...
```
It should take no arguments. For now, just have it "print" something (whatever you want!).

In the future, you'll use the `main()` function to run tests on your code, or to run "scripts" that carry out specific tasks.

Once you add this function and save your file, try running the cell below. You should notice a problem...

In [12]:
# Change this to match your file's name:
import odlib_agrawal as od

od.main()

shyla is the best astronomer ever


**The problem:** After you import a library the first time, Python stops checking it for changes, even if you "import" it again!

**For files you're still working on, you'll need a command called "reload" to update Python with your changes.** The code snippet below tells Python to reload your file.
* Here, we're importing the "reload" command from a built-in library called "importlib", which has tools to help manage "import" statements.

**Try running the cell below.** If your "main()" function works, it should run successfully!

In [13]:
# -- RUN THIS CELL --

from importlib import reload

# Reloads your odlib file:
reload(od)

od.main()

shyla is the best astronomer ever


Congratulations! If you can run the cell above, you've officially started your orbit determination library.

You'll return to this file many times by the end of the program. For now, though, let's move on to some new tools in Python.

<br>

---

# 2) 💫 Loops 💫 

**This section covers:** "For loops", "while loops", and helpful ways to use them

In many programs, it's necessary to complete the same step many times, or even a variable number of times. Examples:
* Given a number "n", print the every number from 1 to n
* Ask a user for a password until the correct password is entered
* Run a video game until a character runs out of health

These situations require you to use a **loop**: a structure that runs the same block of code on repeat until a condition is met.

## **2.1) "For" loops**

In Python, a **for loop** runs a block of code a specific, pre-defined number of times.

Here is a for loop that prints the numbers `0, 1, 2, 3, 4`:
```
for i in range(5):
    # This loop runs once for each value of i in (0, 1, 2, 3, 4)
    print(i)
```

**Syntax of a for loop:**

* Always starts with `for ___ in ____:`

* Like an "if statement", must end with a colon `:` and must be followed by a block of indented code.

* The **range** keyword tells Python how long the loop should be. A loop with `range(5)` will repeat exactly 5 times.

* `i` is known as the "loop variable": it goes up by one each time, and we can use it to keep track of how long the loop has been running.

### **In the cell below, follow along with your instructor to write an example loop:**

In [14]:
# -- WRITE TOGETHER: A simple for loop --

# Write a loop that prints the SQUARE of each number in the range (0, 1, 2, 3, 4)
for i in range(5):
    print(i**2)

0
1
4
9
16


#### **Frequently asked questions about "for loops"**

<br>

**Question 1.** Wait, if we write `range(5)`, why does the loop stop at 4?

**Answer:** Python, like most programming langauges, is ***zero-indexed.*** This means when asked to start counting anything, Python will almost always start counting at $0$.

The function `range(5)` causes Python to count the numbers `0, 1, 2, 3, 4` (a total of 5 numbers).

Zero-indexing can be confusing at first, but you'll get used to it with practice. It also leads to quite a few jokes...

> *A programmer is asked to start a race by counting to three. The programmer nods and says "Zero... one... TWO!"*

<br>

**Question 2.** What if I want to call my loop variable something other than `i`?

**Answer:** You're allowed to call it whatever you want! The following code is just as valid:

```
for tyrannosaurus_rex in range(10):
    # Prints the integers 0, 1, 2... 9
    print(tyrannosaurus_rex)
```

However, `i` is the most common choice.

### **Pause here and write your own `for` loop using a different variable than `i`**

It can do whatever you want, **but keep the "range" below 50!** Printing too many times in a short timespan can cause your notebook to crash.

In [15]:
# -- WRITE YOUR OWN FOR LOOP HERE --

for _ in range(10):
    print("shyla is the best astronomer")

shyla is the best astronomer
shyla is the best astronomer
shyla is the best astronomer
shyla is the best astronomer
shyla is the best astronomer
shyla is the best astronomer
shyla is the best astronomer
shyla is the best astronomer
shyla is the best astronomer
shyla is the best astronomer


## **2.2) Helpful ways to use loops**

Loops are most helpful when combined with other programming techniques.

#### A) Loops and if statements

By using an **"if statement" within a loop,** you can make special things happen for only specific values if `i`

**Example:**

In [16]:
# -- EXAMPLE CODE: An if statement in a loop --

for i in range(6):
    if i == 1:
        print(i, 'is the loneliest number')
    elif i == 3:
        print(i, 'is a party')
    else:
        print(i)

0
1 is the loneliest number
2
3 is a party
4
5


<br>

#### B) Accumulator loops

Another common strategy is the **accumulator loop.** In this type of loop, you create a variable with a certain value, then update it each time the loop runs. This variable *accumulates* changes as the loop runs.

**Example:** Here's a very simple "accumulator loop" that adds the numbers from $0$ to $n$:

In [17]:
# -- EXAMPLE CODE: An accumulator loop --

# Adds the numbers from 0 to n (not including n)
n = 10

total = 0 # Initializes the total to 0
for i in range(n):
    total += i # Updates the total on each loop

print('n =', n, '| sum =', total)

n = 10 | sum = 45


#### C) Nested loops

Finally, you can create **nested loops**, or loops that run inside other loops.
* This is allowed as long as the loop variables have different names! The tradition is to use `i` for the first loop, `j` for the second, `k` for the third, and so on.

The example below uses a nested loop to print every pair of numbers $(i, j)$ between 0 and 5 where $i > j$:

In [18]:
# -- EXAMPLE CODE: Nested loops --

for i in range(5):
    for j in range(i):
        print('i =', i, ' j =', j)
    print('---')

---
i = 1  j = 0
---
i = 2  j = 0
i = 2  j = 1
---
i = 3  j = 0
i = 3  j = 1
i = 3  j = 2
---
i = 4  j = 0
i = 4  j = 1
i = 4  j = 2
i = 4  j = 3
---


<br>

## In-class problems

Using the techniques you've learned, try the problems below on your own!

> #### ❗**Remember: Calling `print(...)` thousands of times in quick succession may crash your notebook!**❗
> 
> As the saying goes, "with great power comes great responsibility". Loops are powerful tools. If write a loop that runs more than a few hundred times, make sure you know what it does!
>
> **If your code is taking too long to run, you can cancel it by pressing the square-shaped "Interrupt the kernel" button at the top of the notebook.**

<br>

### $\rightarrow$ **Problem 2.1:** Spot the loop errors

Each of the cells below has a "for loop" with a serious and common mistake. Find and fix each one.

In [19]:
## -- SPOT THE ERROR --

# Prints the square of each integer from 0 to n
n = 10
for i in range(n):
    print(i, 'squared is', i**2)

0 squared is 0
1 squared is 1
2 squared is 4
3 squared is 9
4 squared is 16
5 squared is 25
6 squared is 36
7 squared is 49
8 squared is 64
9 squared is 81


In [20]:
## -- SPOT THE ERROR --

# Sums the numbers from 0 to 20 (not including 20)
# Prints the running total at each step
running_total = 0
for i in range(20):
    running_total += i
    print(running_total)

0
1
3
6
10
15
21
28
36
45
55
66
78
91
105
120
136
153
171
190


### $\rightarrow$ **Problem 2.2:** Working with loops

In the cell below, write your own code to accomplish as many of the following as you can.

**Loop A)** Print the square of each number from 0 to 9, including 9.

**Loop B)** Print each number from 10 to 1, starting from 10 and ending with 1

**Loop C)** Using an "if" statement, print all whole numbers less than 50 that are divisible by 7.
* Remember, `%` is Python's remainder operator. A number `n` is divisible by 7 if `n % 7 == 0`.

**Loop D)** Use a loop to compute the *factorial* of 5: $5! = 5 \times 4 \times 3 \times 2 \times 1 = 120$

**Loop E)** (Challenge) Print the *number of divisors* of each number from 2 to 30, including 30.
* A whole number $d$ ($1 \leq d < n$) is a *divisor* of $n$ if it divides $n$ with no remainder.
* You'll need to use a nested loop

In [21]:
# -- Problem 2.2 --
print("Loop A)", end=" ")
for i in range(10):
    print(i**2, end=" ")

print("\nLoop B)", end=" ")
for i in range(10, 0, -1):
    print(i, end=" ")

print("\nLoop C)", end=" ")
for i in range(51):
    if (i % 7 == 0):
        print(i, end=" ")

print("\nLoop D)", end=" ")
product = 1
for i in range(1, 6):
    product = product * i
print("5! is", product)

Loop A) 0 1 4 9 16 25 36 49 64 81 
Loop B) 10 9 8 7 6 5 4 3 2 1 
Loop C) 0 7 14 21 28 35 42 49 
Loop D) 5! is 120


<br>

## **2.3) "While" loops**

A *while loop* is a different way to write a loop. It uses the `while` keyword, and it keeps running as long as some condition is `True`:
```
while [CONDITION]:
    # CODE
```

Unlike a "for loop", a "while loop" doesn't to know how many times it will run beforehand! It will run as many times as necessary until its condition is `False`. A while loop is perfect for a situation like this:
```
while password_is_incorrect():
    request_password()
```
Or:
```
while not method_of_gauss_converged():
    improve_orbit()
```

However, also unlike a for loop, a while loop can run *indefinitely*! For instance, the code below will keep running until it's manually stopped. This situation is called an *infinite loop*:
```
# This will keep running forever!
while True:
    do_something()
```
If you're not careful, it's possible (indeed, easy) for your code to get stuck in an infinite loop. Because of that, you need to be more careful when using "while loops" than "for loops".

**Example of a while loop:**

The cell below uses a "while" loop to find the first value of $n$ such that $1 + 1/2 + 1/3 + 1/4 + ... + 1/n > 10$. (This sum is called the [*harmonic series*](https://en.wikipedia.org/wiki/Harmonic_series_(mathematics)).)

We need to use a "while" loop here because we don't know beforehand how many terms we'll need to add before we reach 10.

(Note: If you need to interrupt your code because it's taking too long to run, you can use the square "interrupt this kernel" button at the top of your Jupyter notebook.)

In [22]:
# -- EXAMPLE CODE: A while loop --

n = 0      # Stores the current denominator
total = 0  # Stores the sum of all terms so far

while (total < 10):
    n += 1        # Increases the denominator by 1
    total += 1/n  # Adds 1/n to the total

print('The sum exceeds', 10, 'after', n, 'terms')

The sum exceeds 10 after 12367 terms


<br>

## Before moving on: **Add loops to your reference sheet**

Get out your "Python reference" document and write a simple example of a "for loop" and a "while loop", so that you remember the syntax and can look it up later.

<br>

---

# 3) 📃 Lists 📃

**This section covers:** Creating lists, adding new values to lists, "slicing" lists, and tips for working with lists in practice.

By the end of this section, our goal is to write a function that can **add two vectors together:**
```
def add_vectors(vec1, vec2):
    ???
```

However, there's an immediate problem! So far we've only seen Python variables that store single values. A vector is an object that stores *multiple* values. 

Further, vectors can be any length: an $N$-dimensional vector has $N$ different elements. We'd like our function to work in any number of dimensions. How can we handle this in Python?

**The solution: use a *list* to represent a vector.**

### **3.1) What is a list?**

In Python, a list is an object that stores a sequence of values. It's represented with square brackets. (Notably, this is why you can't use square brackets when doing math.)
```
    list_example = [10, 20, 30]
```

A number's position in this list is called it's ***index***. Like the "range" function, list indices in Python start counting from **0.** (They are "***zero-indexed***"). Here, the number 10 has index 0, 20 has index 1, and 30 has index 2.

To **access** item `i` in a list, you would write `list_example[i]`:
```
    list_example[0]
    -> 10
```



You can **modify** any item in a list like normal:
```
    list_example[0] -= 3
    
    print(list_example)
    -> [7, 20, 30]

    list_example[2] = list_example[0]
    
    print(list_example)
    -> [7, 20, 7]
```

**To get the length of a list, we can use the `len` function:**
```
    len(list_example)
```

The cell below shows a list in action:

In [23]:
# -- EXAMPLE CODE: Creating a list and accessing its values --

# Creates the list
list_example = [10, 15, 30, 60]

print(list_example)
print('Length:', len(list_example))
print('Item 0:', list_example[0])
print('Item 3:', list_example[3])

# Changes the first item
list_example[0] = 5
print('After changing list_example[0], the list is:')
print(list_example)

[10, 15, 30, 60]
Length: 4
Item 0: 10
Item 3: 60
After changing list_example[0], the list is:
[5, 15, 30, 60]


### **3.2) Lists and loops**

We want the function `add_vectors(vec1, vec2)` to work on vectors of any size, which means we can't access the elements in each vector one-by-one, e.g. `vec1[0] + vec2[0]`, `vec1[1] + vec2[1]`, ...

In general, when dealing with lists in Python, you'll want to use a *loop* that iterates over each element in the vector. The standard way to do this is with the following code:
```
    for i in range(len(list_example)):
        # Do something with list_example[i]
```

(The expression `for i in range(len(LIST_NAME)):` is one you'll probably get sick of writing by the end of the program.)

**Example:** The code below uses a for loop to compute the sum of all items in a list. It works for lists of any length, including empty lists!
* The length of an empty list is 0, so the for loop runs 0 times: `for i in range(0)`.

In [24]:
## -- EXAMPLE CODE: Looping over a list --

def get_total(values):
    # Sums the contents of `values`
    total = 0
    for i in range(len(values)):
        total += values[i]
    return total

values = [1, 2, 3, 10, -10, 90]
print(values)
print('This list sums to', get_total(values))
print()

empty = []
print(empty)
print('This list sums to', get_total(empty))

[1, 2, 3, 10, -10, 90]
This list sums to 96

[]
This list sums to 0


### **3.3) Adding values to a list**

You can **add a value to the end of a list with the `append` function**:

```
    print(list_example)
    -> [12, 15, 17]
    
    list_example.append(100)
    
    print(list_example)
    -> [12, 15, 17, 100]
```

A common way to create a list in Python is to start with an empty list, `[]`, and use `append()` to add new items one-by-one.

For example, this code creates a new list that matches `list_example`, but with 10 subtracted from each element:
```
    new_list = []
    for i in range( len(list_example)) ):
        new_list.append( list_example[i] - 10 )
    
    print(new_list)
    -> [2, 5, 7, 90]
```

Additionally, you can **"add" two lists with `+`** to append all of one list's values to the end of another:
```
    new_list = [1, 2, 3] + [4, 5]
    
    print(new_list)
    -> [1, 2, 3, 4, 5]
```

### **3.4) "Slicing" a list**

Index notation (e.g. `list_example[i]`) allows you to look up one element at a time in a list. But what if you want to select multiple elements or a whole section of a list?

Using a ***slice***, you can select many elements at once. "List slicing" is a very versatile tool, and one you'll find yourself using many times! Here is a rundown of how it works:

#### **Taking a slice between two indices**
```
    list_example[start : end]
```
This command selects the set of elements starting from the index **start** and ending just before **end** (NOT including end itself!).

**Example:**
```
    list_example = [0, 10, 20, 30, 40, 50]

    list_example[1 : 4] # Selects the elements at 1, 2, and 3
    -> [10, 20, 30]
```

If you don't provide a "start" or an "end", Python will assume you mean the very start or end of the list.
```
    list_example[: 4] # Selects all elements from the beginning until index 4
    -> [0, 10, 20, 30]
    
    list_example[4 :] # Selects all elements index 4 until the end
    -> [40, 50]
```

#### **Adding a "step size"**
```
    list_example[start : end : step]
```
Optionally, you can add a step size, which will cause the slice to "skip" by `step` indices at a time. You won't always need this, but it's sometimes very helpful.

**Example:**
```
    list_example[1 : 5] # Selects every element from index 1 to 5
    -> [1, 2, 3, 4]
    
    list_example[1 : 5 : 2] # Selects EVERY OTHER element from index 1 to 5
    -> [1, 3]
```
The step size has some unexpected uses! Most notably, **to reverse a list, you can use a slice with step size -1:**
```
    list_example[ : : -1] # Prints every element in reverse order
    -> [5, 4, 3, 2, 1, 0]
```


In [25]:
# -- EXAMPLE CODE: Slicing a list --

list_example = [0, 1, 2, 3, 4, 5]

print(list_example[ : 2]) # Selects all elements up to index 2
print(list_example[2 : ]) # Selects all elements from index 2 til the end
print(list_example[ : : -1]) # Reverses the entire list

[0, 1]
[2, 3, 4, 5]
[5, 4, 3, 2, 1, 0]


### $\rightarrow$ **Mini-exercise:** Creating, adding, and slicing

**Try this on your own!**

Practice using list-related commands by carrying out the steps below. We've given you some example output for comparison.

1. Create an empty list, `append` any 3 items to it, and print it
2. Print the length of the list with `len`
3. "Double the list" by adding all its elements to the end of itself, then print it
4. Drop the first element AND the last 2 elements using a slice, then print it
5. "Reverse the list" by printing its elements in the opposite order, then print it

**Example output:**
```
    [103, 12, -1]                # 1: Create the list & add 3 elements
    3                            # 2: Print the length
    [103, 12, -1, 103, 12, -1]   # 3: Double the list
    [12, -1, 103]                # 4: Drop the first & the last 2
    [103, -1, 12]                # 5: Reverse the list
```

(*Bonus*: What happens when you MULTIPLY a list by a number? By another list?)

In [26]:
# -- Mini-exercise: Create and work with a list -- 
list_1 = [103, 12 ,-1]
print(len(list_1))
list_1 += list_1
print(list_1)
list_1 = list_1[1:4]
print(list_1)
list_1 = list_1[::-1]
print(list_1)

3
[103, 12, -1, 103, 12, -1]
[12, -1, 103]
[103, -1, 12]


<br>

## Before moving on: **Add lists to your reference sheet**

Get out your "Python reference" document and write down some of the list-related commands you'll need to remember!

<br>

### **3.5) Beware! Lists are *mutable***

**What's "mutable"?** *"Mutable"* ("MUTE-able") is a technical term for *"changeable"*. It comes from the same root word as "mutate".

**What does it mean?** Unlike simpler data types, every change made to a list is made directly to that list's location in memory.

This may sound innocent, but list mutability is an extremely common source of bugs, unexpected behaviors, and headaches. To explain why and to help you avoid those, let's take a detailed look.

When working with simple data (numbers, booleans, strings, etc.), when we set one variable equal to another, changes to the new variable don't affect the old one:

In [27]:
# -- EXAMPLE CODE: Using non-lists... --

old_value = 4
new_value = old_value
new_value += 100

print("Old value:", old_value)
print("New value:", new_value)

Old value: 4
New value: 104


***This is not true for lists!*** When you set one list equal to another, changes to the new list also happen to the old one, and vice versa.

In [28]:
# -- EXAMPLE CODE: Using lists... --

old_list = [4]
new_list = old_list
new_list[0] += 100

print("Old list:", old_list)
print("New list:", new_list)

Old list: [104]
New list: [104]


Why does this happen?

`old_list` and `new_list` have different names, but behind-the-scenes, they both refer to the ***exact same location*** in your computer's memory. Since lists can be potentially very large, Python doesn't create a new list in your computer's memory until you ask it to do so.

When you type `print(old_list)` or `print(new_list)`, Python looks up whatever's stored in that spot and prints it. No matter which of these variables you *try* to change, there's only one list in Python's memory to which that change can be applied.

As a result, if you aren't careful, changes to one list can change completely unrelated lists in ways you don't expect!

**The solution: To avoid this issue, make a copy of a list before making changes.**

One way to copy a list is to use the `list` function. Calling `new_list = list(old_list)` causes Python to copy all the values from `old_list` to a new spot in your computer's memory.

```
    old_list = [1,2,3]
    new_list = list(old_list)
    # Now, changes to new_list won't affect old_list (and vice versa)
```

Another way is to build the new list from scratch, starting with an empty list and `appending` new values one-by-one. Which method you use will depend on the situation.

> #### **A realistic example**
>
> Imagine you have your asteroid's current position (a list of its coordinates `[x, y, z]`). You run some code to calculate where it will be in 7 days, and you find that -- according to your code -- the asteroid hasn't moved! The current position and future position are exactly the same.
>
> You check your math backwards and forwards, but nothing seems to be wrong.
> 
> Errors like this often happen when you create a new list from an old one without copying it, like this:
>
> ```
>     future_position = current_position
>     # Changes to future_position
>     
> ```
> After running this code, `future_position` and `current_position` refer to the same list behind the scenes. They appear to be equal because every time your code changed the asteroid's future position, it changed the CURRENT position as well.
>
> You could fix this error by creating a copy of the current position, so changes to the future position won't affect it:
> ```
>     future_position = list(current_position)
>     
> ```
> However, you'll write a lot of code during the program, and it's likely that you'll fo at least once. Remember to check for this type of mistake when you're debugging your code.


In the exercises below, you'll write several functions that take in lists. Be careful that you're always making/modifying a copies, not changing the originals.

For your reference:
```
    def func_name(input_list):

        # This is fine
        new_list = list(input_list)

        # This is also okay
        new_list = []
        for i in range( len(input_list) ):
            new_list.append( input_list[i] )

        # NO: Changes to "new_list" will change "input_list" too
        new_list = input_list

        # ALSO NO: This changes input_list directly
        input_list[0] += 10
```


<br>

## In-class problems

Try the following on your own!

> **One final note!** Never use `list` as a variable's name. This will overwrite the `list` function, which means you won't be able to use it anymore.
>
> If this happens, you can run
>
> ```
>     del list
>     
> ```
> to "delete" your changes and return `list` to its normal behavior.

<br>

### $\rightarrow$ **Problem 3.1:** Working with lists

Implement each of the following functions, ensuring they work by using the test case provided.

**A)** `square_elements`, which returns the input list with all elements squared. The function's comments suggest a structure for you to use.

**B)** `multiply_vector`, which returns the input list times the scalar `mult`.

**C)** `get_differences`, which returns a list of the *difference* between each element and the one preceding it. Note that the final list should be shorter than the original list by 1.

In [29]:
# -- Problem 3.1: square_elements --

def square_elements(values):
    # Returns a new list containing the square of each item in the list "values"

    # SUGGESTED STRUCTURE:
    # Create a new list
    # For each element in "values"...
        # Add this element squared to the new list
    # Return the new list

    # YOUR CODE HERE
    results = []
    for i in values:
        results.append(i**2)
    
    return results

test_case = [1, 2, 3, -10, 0.5]
print('Expected:', [1, 4, 9, 100, 0.25])
print('Actual:  ', square_elements(test_case))

Expected: [1, 4, 9, 100, 0.25]
Actual:   [1, 4, 9, 100, 0.25]


In [30]:
# -- Problem 3.1: multiply_vector --

def multiply_vector(vec, mult):
    # Given a vector "vec" (list) and a multiplier "mult" (float),
    # returns a new list representing the vector "mult * vec"

    # YOUR CODE HERE
    results = []
    for i in vec:
        results.append(i*mult)
    return results

## -- Test case --

test_vec = [10, 15, 20]
test_mult = 3
expected = [30, 45, 60]
print('Expected:', expected)
print('Actual:  ', multiply_vector(test_vec, test_mult))

Expected: [30, 45, 60]
Actual:   [30, 45, 60]


In [31]:
# -- Problem 3.1: get_differences --

def get_differences(values):
    # Returns a list of differences between neighboring items in the list "values"
    
    # YOUR CODE HERE
    results = []
    for i in range(1, len(values)):
        results.append(values[i]-values[i-1])
    return results
## -- Test case --

test_case = [4, 5, 7, 1, 0.5]
expected = [1, 2, -6, -0.5]
print('Expected:', expected)
print('Actual:  ', get_differences(test_case))

Expected: [1, 2, -6, -0.5]
Actual:   [1, 2, -6, -0.5]


### $\rightarrow$ **Problem 3.2:** Add two vectors

You now have the tools you need! Implement the function `add_vectors`.

This function should take two lists `vec1` and `vec2` that represent vectors (you can assume these lists have the same length), then return a new list equal to their vector sum.

Your function should output a new list, and should NOT modify `vec1` or `vec2`.

All we've given you are some test cases. You'll need to write this function from scratch.

In [32]:
# -- Problem 3.2: Add two vectors --

# YOUR CODE HERE
def add_vectors(vec1, vec2):
    results = []
    for i in range(len(vec1)):
        results.append(vec1[i]+vec2[i])
    return results

## -- Test cases --

test_vec1 = [1, 2, 0]
test_vec2 = [0, 0, 3]
expected  = [1, 2, 3]
print('Expected:', expected)
print('Actual:  ', add_vectors(test_vec1, test_vec2))
print()

test_vec1 = [13.0, -11.2, 0.4]
test_vec2 = [-1.1,   4.1, 1.9]
expected  = [11.9,  -7.1, 2.3]
print('Expected:', expected)
print('Actual:  ', add_vectors(test_vec1, test_vec2))
print()

Expected: [1, 2, 3]
Actual:   [1, 2, 3]

Expected: [11.9, -7.1, 2.3]
Actual:   [11.9, -7.1, 2.3]



---

### **Bonus problems:** Try these if you have some extra time!

<br>

**Estimating $\pi$.** The Leibnitz formula for pi states that $\frac{\pi}{4} = 1 - \frac{1}{3} + \frac{1}{5} - \frac{1}{7} + \frac{1}{9} - \frac{1}{11} + ...$ Write a function that computes pi from N terms in this series. How many terms (approximately) are needed to make this estimate accurate to 5 decimal places?

**Guess the number.** The `input()` function allows you to request user input in the middle of a program. The command `number = int(input())` reads a users text input and casts it to an `int`.

Using `input`, along with the code snippet below, write a "guess the number" game. It should generate a random number, then repeatedly ask the user to enter a guess, printing "too high" or "too low" until the number is found. (You can use the keyword `break` to immediately end a loop.)

In [33]:
# -- Bonus problem: Guess the number -- 

from random import randint

random_number = randint(0, 999)

